src/helm/benchmark/scenarios/cleva_scenario.py

```diff
@@ -0,0 +1,1491 @@
+        1. It first maps every entry according to a set of predefined mappings in "verbalizer".
+        2. It then stringifies all fields in the given data point, including processing structured data.
+        3. It finally constructs the input string and reference strings.
+        """
+        # If we define a verbalizer, we map all fields before further processing
+        if templates.get("verbalizer", None) is not None:
+            self._mapping_all(data, templates["verbalizer"])
+
+        if "choices" in data.keys():
+            # We take the corresonding choices and apply the `label` template
+            # Note: we do not allow `label` template to access other fields in multi-choice tasks
+            choices = [self._apply(c, templates.get("label", None), label=c) for c in data["choices"]]
+        # We first convert all fields except `input` to strings
+        transformed_data = copy.deepcopy(data)
+        for k, template in templates.items():
+            if k not in ["input", "verbalizer", "meta", "instruction", "label", "answer_context"]:
+                assert k in data, f"Template key `{k}` is not valid!"
+                transformed_data[k] = self._apply(data[k], template, **data)
+
+        # We then merge all other fields into the `input`
+        data["input"] = templates["input"].format(**transformed_data)
+        if "choices" in data.keys():
+            # Overwrite `choices` to the actual continuations
+            data["choices"] = choices
+        else:
+            # For generation tasks, we allow it to access to other stringified fields
+            kwargs = transformed_data
+            del kwargs["label"]
+            data["label"] = [self._apply(x, templates.get("label", None), **kwargs, label=x) for x in data["label"]]
+        return data
+
+    def _apply(self, data: Any, template: Optional[Any], **kwargs) -> str:
+        """
+        This function constructs a string from the data and template for a given field.
+
+        `data` must have the following type: `str`, `Dict[str, str]`, `List[str]`, `List[Dict[str, str]]`.
+        `template` must have the following type:
+        - `str`: composes a string from all stringified fields including itself (if it is `Dict[str, str]`,
+          it will be flattened out).
+        - `dict`: handle structured data like `List[str]` and `List[Dict[str, str]]` by first obtaining a string
+          for each entry and then combining all strigified entries as the final result.
+        """
+        # If template is a `str`, it composes a string from all fields
+        if isinstance(template, str):
+            # If data is a `Dict[str, str]`, flatten all its key-value pairs and treat them as additional fields
+            if isinstance(data, dict):
+                return template.format(**kwargs, **data)
+            # If data is a `str`, just directly compose the output string from other stringified fields
+            return template.format(**kwargs)
```

Comment:
Is `data` unused in this branch? What's an example of when this happens?

Reply:
Yes. I think the code comment is somewhat misleading and we have revised this. In our implementation, when `template` is a string and `data` is not a dict, it will compose a new string from fields in `kwargs`. `data` will not be used because it does not have a name associated, meaning that we have no idea where `data` should be placed when formatting string with `template`. But it does not mean that `data` could not be accessed by `template`, since `data` and its associated name will be available in `kwargs`. An example is the `input` field ("{review} It is"), which basically composes the prompt "This movie sucks. It is" from fields in `kwargs` (`{"review": "This movie sucks.", "label": [0], "choices": ["negative", "positive"]}`). Another example is the `label` field (" {label}" which adds space in front for language modeling), which does not use `data` but `label` in `kwargs` (`{"label": "How about you?"}`), yet they are the same.